### PR TITLE
feat(router): Validate payment method type in payments request against given payment method data for non-card flows

### PR DESCRIPTION
**Pull request by kritikmodi**
## Type of Change

- [x] Bugfix
- [x] New feature

## Description

Added the following checks for non-card flows :

1. Check if `payment_method_type` is specified.
2. Check if the specified `payment_method_type` corresponds to the parent `payment_method`.
3. Check if the specified `payment_method_data` corresponds to the parent `payment_method`.

## Motivation and Context

This PR resolves [this](https://github.com/juspay/hyperswitch/issues/1102) issue.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code

### CODE REVIEW

**Review by prajjwalkumar17 — CHANGES_REQUESTED** [review]
Add required **Labels** and **Milestone**.

File: crates/router/src/core/payments/helpers.rs [hunk 1]
```diff
@@ -917,6 +917,66 @@ pub(crate) fn validate_payment_method_fields_present(
         },
     )?;
 
+    utils::when(
+        !matches!(req.payment_method, Some(api_enums::PaymentMethod::Card))
```
Comment (vspecky) [inline crates/router/src/core/payments/helpers.rs:921]: I noticed this not but this `matches!()` will also return `false` for when `req.payment_method` is `None` which is basically asserting that even if `payment_method` is not given, `payment_method_type` should be given which is incorrect. The following should fix this afaik
```suggestion
        !matches!(req.payment_method, Some(api_enums::PaymentMethod::Card) | None)
```

**Comment by kritikmodi** [comment]
> Add required **Labels** and **Milestone**.

**Review by vspecky — DISMISSED** [review]

**Review by jarnura — APPROVED** [review]

**Review by prajjwalkumar17 — APPROVED** [review]

**Comment by SanchithHegde** [comment]
1. Please avoid basing your PR on old branches and prefer basing your branches on the `main` branch instead. Having many commits in a PR adds unnecessary noise when debugging.
2. I'll be adding the `M-api-contract-changes` label, since this PR adds additional request validation.

**Review by SanchithHegde — DISMISSED** [review]

**Comment by kritikmodi** [comment]
> * Please avoid basing your PR on old branches and prefer basing your branches on the `main` branch instead. Having many commits in a PR adds unnecessary noise when debugging.
> * I'll be adding the `M-api-contract-changes` label, since this PR adds additional request validation.

Okay sure.

**Review by prajjwalkumar17 — DISMISSED** [review]

**Review by prajjwalkumar17 — DISMISSED** [review]

**Review by jarnura — DISMISSED** [review]

**Review by vspecky — APPROVED** [review]

**Review by prajjwalkumar17 — APPROVED** [review]

**Review by vspecky — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

File: crates/router/src/core/payments/helpers.rs [hunk 2]
```diff
@@ -1362,7 +1363,7 @@ pub(crate) fn validate_payment_method_fields_present(
     req: &api::PaymentsRequest,
 ) -> RouterResult<()> {
     utils::when(
-        req.payment_method.is_none() && req.payment_method_data.is_some(),
+        req.payment_method.is_some() != req.payment_method_data.is_some(),
         || {
             Err(errors::ApiErrorResponse::MissingRequiredField {
                 field_name: "payment_method",
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/payments/helpers.rs:1369]: Please give an appropriate error message here based on the missing value

**Review by prajjwalkumar17 — APPROVED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Pull request merged**